### ISSUE TITLE
SPDK nvmf target somtimes reported error while running SPDK fio_plugin in host machine

### ISSUE THREAD

**Issue by liang2yan** [spdk/spdk#852]
## Expected Behavior
No errors are out.

## Current Behavior
SPDK nvmf target reported following error:
```
rdma.c:3432:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x201bff6c1000 length=131072
rdma.c:3432:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x201bfecc7000 length=131072
rdma.c:3432:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x201bfec33000 length=131072
rdma.c:3432:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x201bffa56000 length=131072
rdma.c:3432:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x201bff7e3000 length=131072
rdma.c:3432:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x201bff7b2000 length=131072
ctrlr.c: 143:spdk_nvmf_ctrlr_keep_alive_poll: *NOTICE*: Disconnecting host from subsystem nqn.2016-06.io.spdk:cnode1 due to keep alive timeout.
rdma.c:3432:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x201bfd744000 length=131072
rdma.c:3432:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x201bfdb79000 length=131072
rdma.c:3432:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x201bfe449000 length=131072
rdma.c:3432:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x201bff034000 length=131072
rdma.c:3432:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x201bfddf2000 length=131072
rdma.c:3432:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x201bfdfd5000 length=131072

```

## Steps to Reproduce
1. run nvmf target with following command:
./app/nvmf_tgt/nvmf_tgt -c nvmf.conf -m 0x3
The nvmf.conf content is as below:
```
[Global]
[Bdev]
[Malloc]
  NumberOfLuns 9
  LunSizeInMB 10240
[Nvmf]
  AcceptorPollRate 10000
[Transport]
  Type RDMA
  MaxQueuesPerSession 128
  MaxQueueDepth 256
[Nvme]
  RetryCount 4
  TimeoutUsec 0
  ActionOnTimeout None
  AdminPollRate 100000
  HotplugEnable No
[Subsystem1]
  NQN nqn.2016-06.io.spdk:cnode1
  Listen RDMA 192.168.120.11:4420
  AllowAnyHost Yes
  Host nqn.2016-06.io.spdk:init
  SN SPDK00000000000001
  MaxNamespaces 20
  Namespace Malloc0
  Namespace Malloc1
  Namespace Malloc2
  Namespace Malloc3
  Namespace Malloc4
```

2. get latest SPDK codes in the host machine, configure the SPDK with fio.
3. Run fio_plugin using following fio job file:
```
[global]
ioengine=spdk
direct=1
verify=0
time_based=1
ramp_time=60
runtime=120
thread=1
numjobs=4
mem_size_mb=40960
bs=1024k
rw=write
iodepth=2
group_reporting=1
[job1]
filename=trtype=RDMA adrfam=IPv4 traddr=192.168.120.11 trsvcid=4420 ns=1
size=8G
[job2]
filename=trtype=RDMA adrfam=IPv4 traddr=192.168.120.11 trsvcid=4420 ns=2
size=8G
[job3]
filename=trtype=RDMA adrfam=IPv4 traddr=192.168.120.11 trsvcid=4420 ns=3
size=8G
[job4]
filename=trtype=RDMA adrfam=IPv4 traddr=192.168.120.11 trsvcid=4420 ns=4
size=8G
[job5]
filename=trtype=RDMA adrfam=IPv4 traddr=192.168.120.11 trsvcid=4420 ns=5
size=8G

```
4.

## Context (Environment including OS version, SPDK version, etc.)
SPDK commit: bcfb2b2b9c5f22bceee4e31d0a191f73adebbe9f

**Comment by changpe1** [comment]
@liang2yan Can you provide more detailed error information from the NVMe-oF side?

**Comment by liang2yan** [comment]
Tried with a little different configuration. 
Run nvmf target with "-m 0xf". Run fio_plugin with "runtime=600". The nvmf target will report same error and the fio_plugin somtimes will crash. The log is as below:

```
(gdb) bt
#0  0x00007f1b8fbe5f2b in raise () from /lib64/libc.so.6
#1  0x00007f1b8fbd0561 in abort () from /lib64/libc.so.6
#2  0x00007f1b8fbd0431 in __assert_fail_base.cold.0 () from /lib64/libc.so.6
#3  0x00007f1b8fbde692 in __assert_fail () from /lib64/libc.so.6
#4  0x00007f1b91a284e3 in nvme_free_request (req=0x2007bfce8780)
    at nvme_internal.h:957
#5  0x00007f1b91a29433 in nvme_qpair_submit_request (qpair=0x3059ee0,
    req=0x2007bfce8780) at nvme_qpair.c:550
#6  0x00007f1b91a22196 in spdk_nvme_ns_cmd_write_with_md (ns=0x2007bffd4860,
    qpair=0x3059ee0, buffer=0x2007bf894000, metadata=0x2007bd444000,
    lba=8577024, lba_count=2048,
    cb_fn=0x7f1b91a15c70 <spdk_fio_completion_cb>, cb_arg=0x7f1b0c000d30,
    io_flags=0, apptag_mask=0, apptag=0) at nvme_ns_cmd.c:750
#7  0x00007f1b91a16146 in spdk_fio_queue (td=0x7f1b6abe8c30,
    io_u=0x7f1b0c000bc0) at fio_plugin.c:811
#8  0x000000000041dae6 in td_io_queue (td=0x7f1b6abe8c30, io_u=0x7f1b0c000bc0)
    at ioengines.c:321
#9  0x000000000046088b in do_io (td=td@entry=0x7f1b6abe8c30,
    bytes_done=bytes_done@entry=0x7f1b4ca90b80) at backend.c:1071
#10 0x0000000000462c18 in thread_main (data=<optimized out>) at backend.c:1774
#11 0x00007f1b90179594 in start_thread () from /lib64/libpthread.so.0
#12 0x00007f1b8fca902f in clone () from /lib64/libc.so.6
```

**Comment by liang2yan** [comment]
Changed the env and fioplugin job file. then this issue can be reproduced every time. The changes are as below:
still run nvmf with same config and "-m 0x3". the fioplugin file is as below:
```
[global]
ioengine=spdk
direct=1
verify=0
time_based=1
ramp_time=60
runtime=600
thread=1
**numjobs=20**
mem_size_mb=40960
bs=1024k
rw=write
**iodepth=128**
group_reporting=1
[job1]
filename=trtype=RDMA adrfam=IPv4 traddr=192.168.123.11 trsvcid=4420 ns=1
size=8G
[job2]
filename=trtype=RDMA adrfam=IPv4 traddr=192.168.123.11 trsvcid=4420 ns=2
size=8G
[job3]
filename=trtype=RDMA adrfam=IPv4 traddr=192.168.123.11 trsvcid=4420 ns=3
size=8G
[job4]
filename=trtype=RDMA adrfam=IPv4 traddr=192.168.123.11 trsvcid=4420 ns=4
size=8G
[job5]
filename=trtype=RDMA adrfam=IPv4 traddr=192.168.123.11 trsvcid=4420 ns=5
size=8G
```
increased the numjobs and iodepth. the nvmf target continue to report error. the log is in the attachment.
[nvmfrunlog.txt](https://github.com/spdk/spdk/files/3395624/nvmfrunlog.txt)

nvmf target reported one memory leak from the log.
The fioplugin crashed everytime.
```

Missing separate debuginfos, use: dnf debuginfo-install glibc-2.27-19.fc28.x86_64
(gdb) bt
#0  0x00007f3e416d6f2b in raise () from /lib64/libc.so.6
#1  0x00007f3e416c1561 in abort () from /lib64/libc.so.6
#2  0x00007f3e416c1431 in __assert_fail_base.cold.0 () from /lib64/libc.so.6
#3  0x00007f3e416cf692 in __assert_fail () from /lib64/libc.so.6
#4  0x00007f3e435195ff in nvme_free_request (req=0x2007bfd72f40)
    at nvme_internal.h:957
#5  0x00007f3e4351a54f in nvme_qpair_submit_request (qpair=0x20e3e50,
    req=0x2007bfd72f40) at nvme_qpair.c:550
#6  0x00007f3e4351325f in spdk_nvme_ns_cmd_write_with_md (ns=0x2007bffd07c0,
    qpair=0x20e3e50, buffer=0x200573635000, metadata=0x0, lba=114688, lba_count=2048,
    cb_fn=0x7f3e43506d20 <spdk_fio_completion_cb>, cb_arg=0x7f3dcc00aaf0, io_flags=0,
    apptag_mask=0, apptag=0) at nvme_ns_cmd.c:750
#7  0x00007f3e43507216 in spdk_fio_queue (td=0x7f3e1c69aed0, io_u=0x7f3dcc00a980)
    at fio_plugin.c:815
#8  0x000000000041dae6 in td_io_queue (td=0x7f3e1c69aed0, io_u=0x7f3dcc00a980)
    at ioengines.c:321
#9  0x000000000046088b in do_io (td=td@entry=0x7f3e1c69aed0,
    bytes_done=bytes_done@entry=0x7f3e02ff4b80) at backend.c:1071
#10 0x0000000000462c18 in thread_main (data=<optimized out>) at backend.c:1774
#11 0x00007f3e41c6a594 in start_thread () from /lib64/libpthread.so.0
#12 0x00007f3e4179a02f in clone () from /lib64/libc.so.6
(gdb)
```

**Comment by changpe1** [comment]
@liang2yan Can you try this patch, https://review.gerrithub.io/c/spdk/spdk/+/461734, it should fix the memory leak issue. Thanks.

**Comment by liang2yan** [comment]
tried this patch, memory leak still existed, it only happened while using ctrl+c to kill nvmf target.

**Comment by changpe1** [comment]
> tried this patch, memory leak still existed, it only happened while using ctrl+c to kill nvmf target.

Good, that's the purpose.  Can you increase the QueueDepth defined in NVMe-oF and try again?

**Comment by liang2yan** [comment]
I have updated following nvmf config:
```
[Transport]
  Type RDMA
  MaxQueuesPerSession 256
  MaxQueueDepth 1024
  MaxIOSize 131072
  IOUnitSize 131072
  InCapsuleDataSize 4096

```
This issue still existed and printed more error messages.

**Comment by liang2yan** [comment]
Used kernel nvmf host library and run fio with same config, SPDK reported same error:
```
rdma.c:3473:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x201bea2dd000 length=131072
rdma.c:3473:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x201bea551000 length=131072
rdma.c:3473:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x201bea8ab000 length=131072
rdma.c:3473:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x201bea383000 length=131072
rdma.c: 885:spdk_nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 16
rdma.c: 885:spdk_nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 16
rdma.c: 885:spdk_nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 16
```
the host machine reported kernel panic, error log is as below:

```
[614355.260594] block nvme0n4: no path available - requeuing I/O
[614355.260639] block nvme0n3: no path available - requeuing I/O
[614355.260729] block nvme0n4: no path available - requeuing I/O
[614355.260739] block nvme0n3: no path available - requeuing I/O
[614355.260844] block nvme0n4: no path available - requeuing I/O
[614355.260864] block nvme0n3: no path available - requeuing I/O
[614355.260959] block nvme0n4: no path available - requeuing I/O
[614355.260983] block nvme0n3: no path available - requeuing I/O
[614355.261102] block nvme0n4: no path available - requeuing I/O
[614355.261130] block nvme0n3: no path available - requeuing I/O
[614357.620885] nvme nvme0: Reconnecting in 10 seconds...
[614367.832835] nvme nvme0: creating 56 I/O queues.
[614368.998053] nvme nvme0: Successfully reconnected (1 attempts)
[614369.040256] nvme0c1n1: detected capacity change from 0 to 10737418240
[614392.082366] nvme_ns_head_make_request: 40870 callbacks suppressed
[614392.082370] block nvme0n1: no path available - requeuing I/O
[614392.082374] block nvme0n1: no path available - requeuing I/O
[614392.082376] block nvme0n1: no path available - requeuing I/O
[614392.082378] block nvme0n1: no path available - requeuing I/O
[614392.082381] block nvme0n1: no path available - requeuing I/O
[614392.082383] block nvme0n1: no path available - requeuing I/O
[614392.082385] block nvme0n1: no path available - requeuing I/O
[614392.082388] block nvme0n1: no path available - requeuing I/O
[614392.082390] block nvme0n1: no path available - requeuing I/O
[614392.082392] block nvme0n1: no path available - requeuing I/O
[614392.095148] nvme0c1n2: detected capacity change from 0 to 10737418240
[614394.113360] nvme nvme0: nvme_report_ns_ids: Identify Descriptors failed
[614394.113364] nvme nvme0: identifiers changed for nsid 3
[614394.120778] nvme nvme0: Reconnecting in 10 seconds...
[614394.137192] block nvme0n3: no path - failing I/O
[614394.137197] block nvme0n3: no path - failing I/O
[614394.137200] block nvme0n3: no path - failing I/O
[614394.137202] block nvme0n3: no path - failing I/O
[614394.137205] block nvme0n3: no path - failing I/O
[614394.137207] block nvme0n3: no path - failing I/O
[614394.137209] block nvme0n3: no path - failing I/O
[614394.137212] block nvme0n3: no path - failing I/O
[614394.137229] block nvme0n3: no path - failing I/O
[614394.137232] block nvme0n3: no path - failing I/O
[614394.138147] nvme nvme0: Identify namespace failed
[614394.168293] Buffer I/O error on dev nvme0n3, logical block 0, async page read
[614394.168302] Buffer I/O error on dev nvme0n3, logical block 0, async page read
[614394.168308] Buffer I/O error on dev nvme0n3, logical block 0, async page read
[614394.168311] ldm_validate_partition_table(): Disk read failed.
[614394.168316] Buffer I/O error on dev nvme0n3, logical block 0, async page read
[614394.168322] Buffer I/O error on dev nvme0n3, logical block 0, async page read
[614394.168328] Buffer I/O error on dev nvme0n3, logical block 0, async page read
[614394.168334] Buffer I/O error on dev nvme0n3, logical block 0, async page read
[614394.168336] Dev nvme0n3: unable to read RDB block 0
[614394.168341] Buffer I/O error on dev nvme0n3, logical block 0, async page read
[614394.168347] Buffer I/O error on dev nvme0n3, logical block 0, async page read
[614394.168349]  nvme0n3: unable to read partition table
[614394.171668] Buffer I/O error on dev nvme0n3, logical block 2621424, async page read
[614394.186215] ldm_validate_partition_table(): Disk read failed.
[614394.186234] Dev nvme0n3: unable to read RDB block 0
[614394.186245]  nvme0n3: unable to read partition table
[614394.186292] BUG: unable to handle kernel NULL pointer dereference at 00000000000000                                                                      00
[614394.187075] #PF error: [WRITE]
[614394.187819] PGD 0 P4D 0
[614394.188513] Oops: 0002 [#1] SMP PTI
[614394.189184] CPU: 49 PID: 1421 Comm: kworker/49:1H Not tainted 5.0.16-100.fc28.x86_6                                                                      4 #1
[614394.189859] Hardware name: Intel Corporation S2600WFT/S2600WFT, BIOS SE5C620.86B.00                                                                      .01.0009.101920170742 10/19/2017
[614394.190537] Workqueue: kblockd nvme_requeue_work [nvme_core]
[614394.191203] RIP: 0010:mempool_free+0x4e/0x80
[614394.191860] Code: ef 5b 5d e9 04 ec 9c 00 c3 48 89 f7 e8 0b d1 75 00 48 63 53 08 3b                                                                       53 04 7d 30 48 8b 4b 10 8d 72 01 48 89 df 89 73 08 48 89 c6 <48> 89 2c d1 e8 f9 ce 75                                                                       00 48 8d 7b 30 31 c9 5b ba 01 00 00 00 be
[614394.193198] RSP: 0018:ffffb7ae07f07da0 EFLAGS: 00010093
[614394.193860] RAX: 0000000000000293 RBX: ffff9fa7462b4678 RCX: 0000000000000000
[614394.194516] RDX: 0000000000000000 RSI: 0000000000000293 RDI: ffff9fa7462b4678
[614394.195169] RBP: ffff9fb1d58e6500 R08: 0000000000000000 R09: ffffffffc0d22d00
[614394.195819] R10: ffff9fb1cefe3600 R11: 0000000000000001 R12: ffff9fb360474000
[614394.196461] R13: 0000000000000001 R14: ffff9fb340b90870 R15: ffff9fb360475a20
[614394.197095] FS:  0000000000000000(0000) GS:ffff9fb48f940000(0000) knlGS:00000000000                                                                      00000
[614394.197737] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[614394.198359] CR2: 0000000000000000 CR3: 00000009f520e001 CR4: 00000000007606e0
[614394.198983] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[614394.199600] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
[614394.200203] PKRU: 55555554
[614394.200792] Call Trace:
[614394.201378]  bio_endio+0x82/0x170
[614394.201958]  nvme_ns_head_make_request+0x1b6/0x1f0 [nvme_core]
[614394.202539]  generic_make_request+0x19e/0x400
[614394.203108]  nvme_requeue_work+0x4e/0x60 [nvme_core]
[614394.203667]  process_one_work+0x1a1/0x3a0
[614394.204217]  worker_thread+0x30/0x380
[614394.204761]  ? process_one_work+0x3a0/0x3a0
[614394.205299]  kthread+0x112/0x130
[614394.205824]  ? kthread_create_on_node+0x60/0x60
[614394.206340]  ret_from_fork+0x35/0x40
[614394.206848] Modules linked in: nvme_rdma nvme_fabrics nvme_core uio_pci_generic uio                                                                       ip6t_rpfilter ip6t_REJECT nf_reject_ipv6 xt_conntrack iscsi_tcp libiscsi_tcp ip_set nf                                                                      netlink ebtable_nat ebtable_broute bridge stp llc ip6table_nat nf_nat_ipv6 ip6table_man                                                                      gle ip6table_raw ip6table_security iptable_nat nf_nat_ipv4 nf_nat nf_conntrack nf_defra                                                                      g_ipv6 nf_defrag_ipv4 iptable_mangle iptable_raw iptable_security ebtable_filter ebtabl                                                                      es ip6table_filter ip6_tables vfat fat ib_isert iscsi_target_mod ib_srpt target_core_mo                                                                      d rpcrdma ib_srp scsi_transport_srp sunrpc rdma_ucm ib_iser intel_rapl rdma_cm skx_edac                                                                       iw_cm ib_umad libiscsi ib_ipoib ib_cm scsi_transport_iscsi nfit x86_pkg_temp_thermal i                                                                      ntel_powerclamp coretemp i40iw kvm mlx5_ib ib_uverbs irqbypass ib_core crct10dif_pclmul                                                                       crc32_pclmul ghash_clmulni_intel intel_cstate intel_uncore intel_rapl_perf mei_me mei                                                                       iTCO_wdt iTCO_vendor_support i2c_i801 ioatdma ipmi_ssif lpc_ich ipmi_si ipmi_devintf ip                                                                      mi_msghandler dca pcc_cpufreq acpi_power_meter
[614394.206876]  acpi_pad xfs libcrc32c mlx5_core ast i2c_algo_bit drm_kms_helper ttm i                                                                      40e drm mlxfw crc32c_intel devlink wmi
[614394.211773] CR2: 0000000000000000
[614394.212323] ---[ end trace 51ab3b51fb3962b1 ]---
[614394.223770] RIP: 0010:mempool_free+0x4e/0x80
[614394.224414] Code: ef 5b 5d e9 04 ec 9c 00 c3 48 89 f7 e8 0b d1 75 00 48 63 53 08 3b                                                                       53 04 7d 30 48 8b 4b 10 8d 72 01 48 89 df 89 73 08 48 89 c6 <48> 89 2c d1 e8 f9 ce 75                                                                       00 48 8d 7b 30 31 c9 5b ba 01 00 00 00 be
[614394.225545] RSP: 0018:ffffb7ae07f07da0 EFLAGS: 00010093
[614394.226068] RAX: 0000000000000293 RBX: ffff9fa7462b4678 RCX: 0000000000000000
[614394.226587] RDX: 0000000000000000 RSI: 0000000000000293 RDI: ffff9fa7462b4678
[614394.227109] RBP: ffff9fb1d58e6500 R08: 0000000000000000 R09: ffffffffc0d22d00
[614394.227634] R10: ffff9fb1cefe3600 R11: 0000000000000001 R12: ffff9fb360474000
[614394.228131] ldm_validate_partition_table(): Disk read failed.
[614394.228167] R13: 0000000000000001 R14: ffff9fb340b90870 R15: ffff9fb360475a20
[614394.228170] FS:  0000000000000000(0000) GS:ffff9fb48f940000(0000) knlGS:00000000000                                                                      00000
[614394.229743] Dev nvme0n3: unable to read RDB block 0
[614394.230307] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[614394.230308] CR2: 0000000000000000 CR3: 00000009f520e001 CR4: 00000000007606e0
[614394.230308] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[614394.230309] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
[614394.230309] PKRU: 55555554
[614394.238179] nvme nvme0: Identify namespace failed

```

**Comment by yidong0635** [comment]
Fio.job configure should using  capital letter:
`[Global]`
using the newly code, I will not allowed you to start.
`[root@waikikibeach49 spdk]# LD_PRELOAD=./examples/nvme/fio_plugin/fio_plugin /usr/src/fio/fio nmicfiojob`
Global: (g=0): rw=read, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=spdk, iodepth=1
...
job1: (g=0): rw=read, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=psync, iodepth=1
job2: (g=0): rw=read, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=psync, iodepth=1
fio-3.0
Starting 4 threads and 2 processes
spdk: must set thread=1 when using spdk plugin
spdk: must set thread=1 when using spdk plugin
spdk: must set thread=1 when using spdk plugin
spdk: must set thread=1 when using spdk plugin
job1: Laying out IO file (1 file / 8192MiB)
job2: Laying out IO file (1 file / 8192MiB)
Jobs: 1 (f=1): [_(4),R(1),_(1)][100.0%][r=20.5MiB/s,w=0KiB/s][r=5248,w=0 IOPS][eta 00m:00s]
job1: (groupid=0, jobs=1): err= 0: pid=29750: Mon Sep  2 12:17:34 2019
   read: IOPS=6894, BW=26.9MiB/s (28.2MB/s)(8192MiB/304170msec)
    clat (nsec): min=700, max=315110k, avg=144458.90, stdev=2146114.86
     lat (nsec): min=719, max=315110k, avg=144531.77, stdev=2146114.26
    clat percentiles (nsec):
     |  1.00th=[     940],  5.00th=[    1352], 10.00th=[    1592],
     | 20.00th=[    2040], 30.00th=[    2064], 40.00th=[    2096],
     | 50.00th=[    2128], 60.00th=[    2160], 70.00th=[    2288],
     | 80.00th=[    2448], 90.00th=[    2544], 95.00th=[    2736],
     | 99.00th=[  741376], 99.50th=[10158080], 99.90th=[35389440],
     | 99.95th=[43253760], 99.99th=[51642368]
   bw (  KiB/s): min= 6144, max=39424, per=49.97%, avg=27561.64, stdev=3794.08, samples=608
   iops        : min= 1536, max= 9856, avg=6890.40, stdev=948.53, samples=608
  lat (nsec)   : 750=0.01%, 1000=1.23%
  lat (usec)   : 2=13.81%, 4=82.29%, 10=0.68%, 20=0.31%, 50=0.11%
  lat (usec)   : 100=0.01%, 250=0.01%, 500=0.49%, 750=0.07%, 1000=0.30%
  lat (msec)   : 2=0.12%, 4=0.03%, 10=0.03%, 20=0.23%, 50=0.27%
  lat (msec)   : 100=0.02%, 250=0.01%, 500=0.01%
  cpu          : usr=0.74%, sys=2.11%, ctx=32955, majf=0, minf=14
  IO depths    : 1=100.0%, 2=0.0%, 4=0.0%, 8=0.0%, 16=0.0%, 32=0.0%, >=64=0.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     issued rwt: total=2097152,0,0, short=0,0,0, dropped=0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=1
job2: (groupid=0, jobs=1): err= 0: pid=29751: Mon Sep  2 12:17:34 2019
   read: IOPS=6931, BW=27.1MiB/s (28.4MB/s)(8192MiB/302534msec)
    clat (nsec): min=570, max=351773k, avg=143681.47, stdev=2147579.81
     lat (nsec): min=589, max=351773k, avg=143754.36, stdev=2147579.17
    clat percentiles (nsec):
     |  1.00th=[     932],  5.00th=[    1352], 10.00th=[    1576],
     | 20.00th=[    2024], 30.00th=[    2064], 40.00th=[    2096],
     | 50.00th=[    2128], 60.00th=[    2192], 70.00th=[    2320],
     | 80.00th=[    2448], 90.00th=[    2544], 95.00th=[    2736],
     | 99.00th=[  741376], 99.50th=[10158080], 99.90th=[35389440],
     | 99.95th=[42729472], 99.99th=[51642368]
   bw (  KiB/s): min= 8704, max=41389, per=50.27%, avg=27724.90, stdev=3846.74, samples=605
   iops        : min= 2176, max=10347, avg=6931.22, stdev=961.69, samples=605
  lat (nsec)   : 750=0.01%, 1000=1.27%
  lat (usec)   : 2=14.55%, 4=81.49%, 10=0.68%, 20=0.32%, 50=0.11%
  lat (usec)   : 100=0.01%, 250=0.01%, 500=0.49%, 750=0.07%, 1000=0.30%
  lat (msec)   : 2=0.12%, 4=0.04%, 10=0.02%, 20=0.23%, 50=0.27%
  lat (msec)   : 100=0.02%, 250=0.01%, 500=0.01%
  cpu          : usr=0.72%, sys=2.14%, ctx=32941, majf=0, minf=20
  IO depths    : 1=100.0%, 2=0.0%, 4=0.0%, 8=0.0%, 16=0.0%, 32=0.0%, >=64=0.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     issued rwt: total=2097152,0,0, short=0,0,0, dropped=0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=1

Run status group 0 (all jobs):
   READ: bw=53.9MiB/s (56.5MB/s), 26.9MiB/s-27.1MiB/s (28.2MB/s-28.4MB/s), io=16.0GiB (17.2GB), run=302534-304170msec

Disk stats (read/write):
  sda: ios=66678/2005, merge=34/3372, ticks=1302246/10480, in_queue=1277227, util=20.94%

> [root@waikikibeach40 spdk]# ./app/nvmf_tgt/nvmf_tgt -c ../nvmf.conf

rpc.c: 277:spdk_rpc_register_alias_deprecated: *ERROR*: cannot create alias bdev_nvme_set_options - method set_bdev_nvme_options does not exist
rpc.c: 277:spdk_rpc_register_alias_deprecated: *ERROR*: cannot create alias bdev_nvme_set_hotplug - method set_bdev_nvme_hotplug does not exist
Starting SPDK v19.10-pre / DPDK 19.05.0 initialization...
[ DPDK EAL parameters: nvmf --no-shconf -c 0x1 --log-level=lib.eal:6 --log-level=lib.cryptodev:5 --log-level=user1:6 --base-virtaddr=0x200000000000 --match-allocations --file-prefix=spdk_pid120333 ]
app.c: 630:spdk_app_start: *NOTICE*: Total cores available: 1
reactor.c: 337:_spdk_reactor_run: *NOTICE*: Reactor started on core 0
conf.c: 136:spdk_nvmf_read_config_file_tgt_conf: *NOTICE*: The value of ConnectionScheduler is not configured,
we will use RoundRobin as the default scheduler
conf.c: 318:spdk_nvmf_parse_subsystem: *NOTICE*: Subsystem nqn.2016-06.io.spdk:cnode1: missing model number, will use default

> From now on, can't duplicate this issue.

**Comment by shadowhzx** [comment]
I have finished the  automation of using script of run_nvmf.py to test this case.
using fio_plugin through bdev layer, the path is **examples/bdev/fio_plugin/fio_plugin**:
the test result is : 
when the io size >= 4k, it  happens:

```
rdma.c:3448:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x20001d803000 length=6144
rdma.c:3394:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x20001d803000 length=6144
rdma.c:3448:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x20001849b000 length=6144
rdma.c:3394:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x20001849b000 length=6144
rdma.c:3448:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x20001c819000 length=6144
rdma.c:3394:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x20001c819000 length=6144
```
the config.json file :
```json
{
  "general": {
    "username": "root",
    "password": "tester",
    "transport": "rdma"
  },
  "target": {
    "nic_ips": ["192.168.89.11"],
    "mode": "spdk",
    "use_null_block": false,
    "nvmet_dir": "/path/to/nvmetcli",
    "num_cores": "1",
    "num_shared_buffers": 8192
  },
  "initiator1": {
    "ip": "10.67.110.204",
    "nic_ips": ["192.168.89.11"],
    "mode": "spdk"
  },
  "fio": {
    "bs": ["6k"],
    "qd": [128],
    "rw": ["randrw"],
    "rwmixread": 100,
    "run_time": 10,
    "ramp_time": 1,
    "run_num": 3,
    "num_jobs": 3
  }
}

```

**Comment by changpe1** [comment]
@yidong0635 I guess the issue is still exist, you can use large IO size for test, e.g. 1024KiB.

**Comment by yidong0635** [comment]
@changpe1  we use scripts to realize the auto testing of this issue. With the new code recovery mechanism from Seth,  we decide to close this issue , and submit a new issue.

**Comment by shadowhzx** [comment]
when running the nvme filesystem with two initiator , this issue is reproducible

**Comment by liang2yan** [comment]
Run nvmf 1 target and 2 hosts file system testing, this issue will be reproduced. Following messages are out:
```
rdma.c:3836:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x20006971f000 length=131072
rdma.c:3765:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x20006a12f000 length=131072
rdma.c:3836:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x20006a831000 length=131072
rdma.c:3836:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x200069503000 length=131072
rdma.c:3765:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x20006a9b3000 length=131072
rdma.c:3836:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x20006a524000 length=131072
rdma.c:3836:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x20006a53f000 length=131072
rdma.c: 919:spdk_nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 16
rdma.c:3765:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x20006a6e8000 length=32768
transport.c: 150:spdk_nvmf_transport_destroy: *ERROR*: transport buffer pool count is 3344 but should be 4096
```

**Comment by sashakot** [comment]
Hi,

Can you add details about your environment?
CPU arch, OS, SW stack (MOFED/inbox), NIC type  

BR
Sasha

**Comment by liang2yan** [comment]
Hi,

The env details are as below.
target:
CPU - skylake
OS - CentOS 7.7.1908 5.2.0-1.el7.elrepo.x86_64
SW stack - rdma-core-22.1-3.el7.src.rpm
NIC - Mellanox MCX416A-CCAT  
Host 1:
CPU - Broadwell
OS - CentOS 7.4.1708 5.2.3-1.el7.elrepo.x86_64
SW Stack - rdma-core-17.2-3.el7.src.rpm
NIC type - MCX516A-CDAT
Host 2:
CPU - Broadwell
OS - CentOS 7.4.1708 5.2.11-1.el7.elrepo.x86_64
SW Stack - rdma-core-17.2-3.el7.src.rpm
NIC Type - MCX456A-ECAT

**Comment by liang2yan** [comment]
@AlekseyMarchuk , do you have any update about this issue? the issue 1086 can't be reproduced and closed.

**Comment by jimharris** [comment]
[Bug Scrub] @AlekseyMarchuk - can I assign this issue to you for further investigation?  @liang2yan says this issue can still be reproduced.

**Comment by AlekseyMarchuk** [comment]
@jimharris sure, I'll check this issue

**Comment by AlekseyMarchuk** [comment]
Hi @liang2yan 

I tried to reproduce this issue about 20 times and I didn't see any errors on target/initiator side. I used spdk version `12ae64f86` and the following target configuration copied from your config:
```
[Malloc]
  NumberOfLuns 5
  LunSizeInMB 10240
[Nvmf]
  AcceptorPollRate 10000
[Transport]
  Type RDMA
  MaxQueuesPerSession 256
  MaxQueueDepth 1024
  MaxIOSize 131072
  IOUnitSize 131072
  InCapsuleDataSize 4096
[Nvme]
  RetryCount 4
  TimeoutUsec 0
  ActionOnTimeout None
  AdminPollRate 100000
  HotplugEnable No
[Subsystem1]
  NQN nqn.2016-06.io.spdk:cnode1
  Listen RDMA 192.168.1.1:4420
  AllowAnyHost Yes
  Host nqn.2016-06.io.spdk:init
  SN SPDK00000000000001
  MaxNamespaces 20
  Namespace Malloc0
  Namespace Malloc1
  Namespace Malloc2
  Namespace Malloc3
  Namespace Malloc4
```
fio config:
```
[global]
ioengine=spdk
norandommap=1
direct=1
verify=0
time_based=1
ramp_time=60
runtime=120
thread=1
numjobs=20
bs=1024k
rw=write
iodepth=128
group_reporting=1
cpus_allowed=0-27
[job1]
filename=trtype=RDMA adrfam=IPv4 traddr=192.168.1.1 trsvcid=4420 ns=1
size=8G
[job2]
filename=trtype=RDMA adrfam=IPv4 traddr=192.168.1.1 trsvcid=4420 ns=2
size=8G
[job3]
filename=trtype=RDMA adrfam=IPv4 traddr=192.168.1.1 trsvcid=4420 ns=3
size=8G
[job4]
filename=trtype=RDMA adrfam=IPv4 traddr=192.168.1.1 trsvcid=4420 ns=4
size=8G
[job5]
filename=trtype=RDMA adrfam=IPv4 traddr=192.168.1.1 trsvcid=4420 ns=5
size=8G
```

Could you clarify if you still see this error?

**Comment by AlekseyMarchuk** [comment]
@liang2yan I've just remembered that I noticed such errors in SPDK target in the case when I killed SPDK initiator, and as far as I remember, upgrade of in-box IB drivers resolved this issue. I was trying to reproduce this issue with MLNX_OFED_LINUX-4.6-1.0.1.1. rdma-core library in MOFED differs from the upstream version also MOFED upgrades FW on NIC. If this issue is still reproducible on your setup, will it be possible for you to re-check it with the latest MOFED?

**Comment by jgrodman** [comment]
I am also getting this error running a malloc bdev over RDMA

I am running nvmf_tgt with 1 core: `./nvmf_tgt -m 0x1`
I am running a single malloc bdev and exposing it over RDMA - see attached json file that I am loading via `./rpc.py load_config < rdma_malloc.json` (attached here as txt file because of github restriction)
[rdma_malloc.txt](https://github.com/spdk/spdk/files/4165097/rdma_malloc.txt)


Then I connect from the initiator via nvme-cli: `nvme connect-all -t rdma -a [IP] -s 4420`
And I run fio with the following job file:
```
[global]
rw=read
ioengine=libaio
blocksize=1M
iodepth=1000
numjobs=40
direct=1
runtime=30
time_based

[queryA0]
filename=/dev/nvme0n1
```


I forked spdk and the most recent commit I have is:
```
commit ae3a9b8f08de94e95f6ee700d4901903bc898bd9
Author: Seth Howell <seth.howell@intel.com>
Date:   Tue Oct 15 12:43:30 2019 -0700
```

On the nvmf_tgt I see the following errors:

```
Feb  6 12:23:03 s1-76116 spdk[31735]: rdma.c:3836:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x200201a30000 length=131072
Feb  6 12:23:03 s1-76116 spdk[31735]: rdma.c:3765:spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x200201a30000 length=131072
```


On the initiator I see the following errors, and the initiator throughput also drops to 0:
```
fio: io_u error on file /dev/nvme0n1: Input/output error: read offset=17825792, buflen=1048576
fio: pid=6355, err=5/file:io_u.c:1756, func=io_u error, error=Input/output error
```

**Comment by AlekseyMarchuk** [comment]
Hi @jgrodman 

I ran fio with your configuration about 30 times and didn't notice this error. I used RHEL 7.4 3.10.0-693.el7.x86_64 with MLNX_OFED_LINUX-4.6-1.0.1.1 on both machines. Could you clarify your environment? I think that the issue is caused by inbox drivers (rdma-core)

**Comment by jgrodman** [comment]
Hi @AlekseyMarchuk thanks for giving it a look...here is my setup:

target and initators:

```
# uname -mrsn
Linux s1-76116 4.15.0-52-generic x86_64
```

```
# cat /etc/os-release
NAME="Ubuntu"
VERSION="18.04.2 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.2 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```

target:
```
# lshw -class network
  *-network
       description: Ethernet interface
       product: MT27520 Family [ConnectX-3 Pro]
       vendor: Mellanox Technologies
       physical id: 0
       bus info: pci@0000:82:00.0
       logical name: Data
       version: 00
       serial: 7c:fe:90:93:43:01
       width: 64 bits
       clock: 33MHz
       capabilities: pm vpd msix pciexpress bus_master cap_list rom ethernet physical fibre autonegotiation
       configuration: autonegotiation=on broadcast=yes driver=mlx4_en driverversion=4.0-0 duplex=full firmware=2.34.5000 ip=<IP> latency=0 link=yes multicast=yes
       resources: irq:65 memory:fbe00000-fbefffff memory:fb000000-fb7fffff memory:fbd00000-fbdfffff memory:f7000000-faffffff
  *-network
       description: Ethernet interface
       physical id: 2
       bus info: pci@0000:82:00.0
       logical name: so00
       serial: 7c:fe:90:93:43:00
       capabilities: ethernet physical fibre autonegotiation
       configuration: autonegotiation=off broadcast=yes driver=mlx4_en driverversion=4.0-0 firmware=2.34.5000 link=no multicast=yes port=fibre
```


initiator 1 and 2:
```
# apt list --installed rdma-core
Listing... Done
rdma-core/bionic-updates,now 17.1-1ubuntu0.1 amd64 [installed]
```

initator 1:
```
# lshw -class network
  *-network
       description: Ethernet interface
       product: MT27520 Family [ConnectX-3 Pro]
       vendor: Mellanox Technologies
       physical id: 0
       bus info: pci@0000:04:00.0
       logical name: so00
       version: 00
       serial: 50:6b:4b:69:9a:00
       width: 64 bits
       clock: 33MHz
       capabilities: pm vpd msix pciexpress bus_master cap_list rom ethernet physical fibre autonegotiation
       configuration: autonegotiation=off broadcast=yes driver=mlx4_en driverversion=4.0-0 firmware=2.40.5048 latency=0 link=no multicast=yes port=fibre
       resources: irq:47 memory:d9f00000-d9ffffff memory:d3800000-d3ffffff memory:d9000000-d90fffff memory:cc000000-cfffffff
  *-network
       description: Ethernet interface
       physical id: 1
       bus info: pci@0000:04:00.0
       logical name: Data
       serial: 50:6b:4b:69:9a:01
       capabilities: ethernet physical fibre autonegotiation
       configuration: autonegotiation=on broadcast=yes driver=mlx4_en driverversion=4.0-0 duplex=full firmware=2.40.5048 ip=<IP> link=yes multicast=yes
```

initiator 2:
```
# lshw -class network
  *-network
       description: Ethernet interface
       product: MT27520 Family [ConnectX-3 Pro]
       vendor: Mellanox Technologies
       physical id: 0
       bus info: pci@0000:24:00.0
       logical name: so01
       version: 00
       serial: ec:0d:9a:16:18:01
       width: 64 bits
       clock: 33MHz
       capabilities: pm vpd msix pciexpress bus_master cap_list ethernet physical fibre autonegotiation
       configuration: autonegotiation=off broadcast=yes driver=mlx4_en driverversion=4.0-0 firmware=2.40.5030 latency=0 link=no multicast=yes port=fibre
       resources: irq:95 memory:fbf00000-fbffffff memory:fb000000-fb7fffff memory:f7000000-faffffff
  *-network
       description: Ethernet interface
       physical id: 3
       bus info: pci@0000:24:00.0
       logical name: Data
       serial: ec:0d:9a:16:18:00
       capabilities: ethernet physical fibre autonegotiation
       configuration: autonegotiation=on broadcast=yes driver=mlx4_en driverversion=4.0-0 duplex=full firmware=2.40.5030 ip=<IP> link=yes multicast=yes
```

**Comment by AlekseyMarchuk** [comment]
I'll find Ubuntu 18.04 to check my assumption regarding inbox drivers. @jgrodman will you be able to install the latest release of OFED/MOFED and check that these errors still exist?

**Comment by AlekseyMarchuk** [comment]
I checked Ubuntu 18.04 with standard ib/rdmacm drivers and didn't see these errors. @jgrodman if it is possible for you to install MOFED, I would wait for your results since MOFED installation also updates FW on NIC.

**Comment by jgrodman** [comment]
@AlekseyMarchuk I am going to try to reproduct this on the most recent version of ubuntu and see if the error persists...if it does then I will try OFED/MOFED

**Comment by yidong0635** [comment]
@AlekseyMarchuk,@Seth5141 ,@benlwalker,@changpe1, @jimharris ,@liang2yan ,
since I have the real environment to simulate this issue, and I add print to debug this, found that something valuable:
1. this issue occurred at test pool before.
2.` spdk_nvmf_rdma_poller_poll: *ERROR*: data=0x200201a30000 length=131072 `
it shows not only in RDMA_WR_TYPE_SEND but also  in case RDMA_WR_TYPE_DATA,  the error message are the same.
and the error message is  5:
`IBV_WC_WR_FLUSH_ERR (5) - Work Request Flushed Error`: A Work Request was in process or outstanding when the QP transitioned into the Error State.
that means some command was blocked by QueuePair into a error state, then the left commands would keep IBV_WC_WR_FLUSH_ERR  .
and I observed that :
`rqpair->current_send_depth -= rdma_req->num_outstanding_data_wr+1;`
3. If you search the code, there are several codes about calculating rqpair->current_send_depth,
maybe there are some wrong calculations for a  queue sequence . Which leads IBV_WC_WR_FLUSH_ERR .

 I'm no sure about this, but test verified this.
thanks!

**Comment by AlekseyMarchuk** [comment]
Hi @yidong0635 

Could you clarify your environment, I want to reproduce the issue on my side. I wonder why I could not reproduce it for so long..

**Comment by yidong0635** [comment]
[perf.log](https://github.com/spdk/spdk/files/4286885/perf.log) the important is the fio config and testing methods: the environment is provided by Liang.
there's one target, three initiators:
target:
5.4.8-1.el7.elrepo.x86_64 #1 SMP Sat Jan 4 15:29:03 EST 2020 x86_64 x86_64 x86_64 GNU/Linux

we use json config:
{
  "general": {
    "username": "***",
    "password": " ***",
    "transport": "rdma"
  },
  "target": {
    "nic_ips": ["192.168.35.11"],
    "mode": "spdk",
    "use_null_block": false,
    "nvmet_dir": "/path/to/nvmetcli",
    "num_cores": "1",
    "num_shared_buffers": 8192
  },
  "initiator1": {
    "ip": "10.67.110.195",
    "nic_ips": ["192.168.35.11"],
    "mode": "kernel",
    "nvmecli_dir": "/usr/sbin",
    "fio_dir": "/usr/src/fio"
  },
  "initiator2": {
    "ip": "10.67.111.120",
    "nic_ips": ["192.168.35.11"],
    "mode": "kernel",
    "nvmecli_dir": "/usr/sbin",
    "fio_dir": "/usr/src/fio"
  },
  "initiator3": {
    "ip": "10.67.111.151",
    "nic_ips": ["192.168.35.11"],
    "mode": "kernel",
    "fio_dir": "/usr/src/fio",
    "nvmecli_dir": "/usr/local/sbin"
  },
  "fio": {
    **"bs": ["1048576"],
    "qd": [64],
    "rw": ["randrw"],
    "rwmixread": 60,
    "run_time": 72,**
    "ramp_time": 1,
    "run_num": 1,
    "num_jobs": 1
  }
 it can be reproduced every time, change code can avoid this error .

**Comment by AlekseyMarchuk** [comment]
@yidong0635 Thanks for your reply. Could you clarify what you mean by `change code can avoid this error`?

**Comment by yidong0635** [comment]
@AlekseyMarchuk  I know that 
[fio.log](https://github.com/spdk/spdk/files/4290253/fio.log)
maybe may change is not right, just to verify this is not due to FW/OS/OFED/MOFED:
```
diff --git a/lib/nvmf/rdma.c b/lib/nvmf/rdma.c
index a31c9ba..d73d2ec 100644
--- a/lib/nvmf/rdma.c
+++ b/lib/nvmf/rdma.c
@@ -2242,7 +2242,7 @@ spdk_nvmf_rdma_request_process(struct spdk_nvmf_rdma_transport *rtransport,
                                /* This request needs to wait in line to perform RDMA */
                                break;
                        }
-                       if ((rqpair->current_send_depth + rdma_req->num_outstanding_data_wr + 1) >
+                       if ((rqpair->current_send_depth + rdma_req->num_outstanding_data_wr) >
                            rqpair->max_send_depth) {
                                /* We can only have so many WRs outstanding. we have to wait until some finish.
                                 * +1 since each request has an additional wr in the resp. */
@@ -3873,7 +3873,7 @@ spdk_nvmf_rdma_poller_poll(struct spdk_nvmf_rdma_transport *rtransport,

                        rdma_req->state = RDMA_REQUEST_STATE_COMPLETED;
                        /* +1 for the response wr */
-                       rqpair->current_send_depth -= rdma_req->num_outstanding_data_wr + 1;
+                       rqpair->current_send_depth -= rdma_req->num_outstanding_data_wr;
                        rdma_req->num_outstanding_data_wr = 0;

                        spdk_nvmf_rdma_request_process(rtransport, rdma_req);
```
it isn't totally right, _**_but this can see obvious effect_**_.
I feel the calculation of the length is not right this would due to IBV_WC_WR_FLUSH_ERR .

**Comment by AlekseyMarchuk** [comment]
Hi @yidong0635 
IBV_WC_WR_FLUSH_ERR errors are generated for all work requests after qpair moves to an error state. Errors which represent send/receive queue overflow are IBV_WC_LOC_LEN_ERR and IBV_WC_REM_INV_REQ_ERR, have you seen such errors?
I'll check your assumption regarding wrong calculation of send depth, but I believe it should be correct, otherwise we'll see a lot of errors in our performance tests
Could you also clarify your distro version?

**Comment by yidong0635** [comment]
@AlekseyMarchuk , you are right, that's my assumption. My distro is to download the latest version every day.
And I applied your patch on gerrit. It shows retried counter exceeded (12) then work flush error(5).
My change before may changed the rsp which I tested the rwmixread": 60%-->1%", it reduced the error probability, this is experience value,It's not right, please ignore.
rwmixread=1 which means write(99%) and read(1%).  It indicates fio write got the error.

**Comment by AlekseyMarchuk** [comment]
@yidong0635 This error says that something is wrong with connectivity between your nodes. I may seem stubborn, but could you try to install the latest ofed/mofed on your nodes? Sometimes in-box drivers (e.g. rdma-core) have their own forks that don't include all the latest fixes.

**Comment by yidong0635** [comment]
@AlekseyMarchuk  we are finding the right ofed/mofeds. Will do this right now.

**Comment by yidong0635** [comment]
@AlekseyMarchuk  after installing the right ofed/mofeds, due to the code update of error reporting,
errors are changed: 
paste some of them
[perf.txt](https://github.com/spdk/spdk/files/4380336/perf.txt)

```
rdma.c:3950:spdk_nvmf_rdma_poller_poll: *ERROR*: Error on CQ 0x1ed18f0, request 0x139913354541248, type 1, status: (5): Work Request Flushed Error
rdma.c:3950:spdk_nvmf_rdma_poller_poll: *ERROR*: Error on CQ 0x1ed18f0, request 0x139913354496624, type 2, status: (5): Work Request Flushed Error
rdma.c:3950:spdk_nvmf_rdma_poller_poll: *ERROR*: Error on CQ 0x1ed18f0, request 0x139913354496472, type 1, status: (5): Work Request Flushed Error
rdma.c:3950:spdk_nvmf_rdma_poller_poll: *ERROR*: Error on CQ 0x1ed18f0, request 0x139913354513608, type 2, status: (5): Work Request Flushed Error
rdma.c:3950:spdk_nvmf_rdma_poller_poll: *ERROR*: Error on CQ 0x1ed18f0, request 0x139913354513456, type 1, status: (5): Work Request Flushed Error
rdma.c:3950:spdk_nvmf_rdma_poller_poll: *ERROR*: Error on CQ 0x1ed18f0, request 0x139913354527504, type 2, status: (5): Work Request Flushed Error
rdma.c:3950:spdk_nvmf_rdma_poller_poll: *ERROR*: Error on CQ 0x1ed18f0, request 0x139913354527352, type 1, status: (5): Work Request Flushed Error

rdma.c:3950:spdk_nvmf_rdma_poller_poll: *ERROR*: Error on CQ 0x1ed18f0, request 0x139913354527352, type 1, status: (5): Work Request Flushed Error
rdma.c: 942:spdk_nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 16
rdma.c: 942:spdk_nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 10
[initiator2]
[initiator2]
[initiator2] FIO run finished. Results in: /usr/src/fio/nvmf_perf/1048576_64_randrw_m_30_run_1_initiator2.json
rdma.c: 942:spdk_nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 16
rdma.c: 942:spdk_nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 16
rdma.c: 942:spdk_nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 16
rdma.c: 942:spdk_nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 7
rdma.c: 942:spdk_nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 16
rdma.c: 942:spdk_nvmf_rdma_qpair_destroy: *WARNING*: Destroying qpair when queue depth is 16
[initiator3]

```

**Comment by AlekseyMarchuk** [comment]
@yidong0635 I suggest excluding connectivity issues between your nodes. Could you try to run ib_write_bw tool (part of MOFED) on target and initiator nodes and check if it reports any errors?
On my side the commands look like this:
Target: `sudo ib_write_bw --rdma_cm --bidirectional --all --ib-dev=mlx5_1 --run_infinitely`
Initiator: `sudo ib_write_bw --rdma_cm --bidirectional --all --ib-dev=mlx5_1 --run_infinitely target_ib_dev_ip`
Let both processes run for several minutes and if no errors reported kill them with ctrl+c
Thank you!

**Comment by yidong0635** [comment]
@AlekseyMarchuk 
Ok, I will try to run this commands:
and compared with the configs :
the key factor is "bs": ["1048576"] (reports above error )vs "bs": ["4096"] ( no error),  rwmixread : 50, (write causes error, so "0" is more likely )

**Comment by yidong0635** [comment]
@AlekseyMarchuk  use this two machine (one target and one client with above config) can  also reproduce this issue. And the following is the 
`"ib_write_bw --rdma_cm --bidirectional --all --ib-dev=mlx5_1 --run_infinitely"` log
```
sudo ib_write_bw --rdma_cm --bidirectional --all --ib-dev=mlx5_1 --run_infinitely

************************************
* Waiting for client to connect... *
************************************
---------------------------------------------------------------------------------------
                    RDMA_Write Bidirectional BW Test
 Dual-port       : OFF          Device         : mlx5_1
 Number of qps   : 1            Transport type : IB
 Connection type : RC           Using SRQ      : OFF
 PCIe relax order: ON
 TX depth        : 128
 CQ Moderation   : 1
 Mtu             : 1024[B]
 Link type       : Ethernet
 GID index       : 2
 Max inline data : 0[B]
 rdma_cm QPs     : ON
 Data ex. method : rdma_cm
---------------------------------------------------------------------------------------
 Waiting for client rdma_cm QP to connect
 Please run the same command with the IB/RoCE interface IP
---------------------------------------------------------------------------------------
 local address: LID 0000 QPN 0x57a8 PSN 0x9e92eb
 GID: 00:00:00:00:00:00:00:00:00:00:255:255:192:168:35:11
 remote address: LID 0000 QPN 0x1306 PSN 0x960375
 GID: 00:00:00:00:00:00:00:00:00:00:255:255:192:168:35:155
---------------------------------------------------------------------------------------
 #bytes     #iterations    BW peak[MB/sec]    BW average[MB/sec]   MsgRate[Mpps]
Conflicting CPU frequency values detected: 3628.994000 != 999.888000. CPU Frequency is not max.
 65536      88278            0.00               2205.48            0.035288
Conflicting CPU frequency values detected: 3630.434000 != 1000.199000. CPU Frequency is not max.
 65536      88278            0.00               2206.92            0.035311
Conflicting CPU frequency values detected: 3632.907000 != 1000.343000. CPU Frequency is not max.
 65536      88280            0.00               2206.92            0.035311
Conflicting CPU frequency values detected: 3611.730000 != 1000.179000. CPU Frequency is not max.
 65536      88279            0.00               2206.94            0.035311
Conflicting CPU frequency values detected: 3630.561000 != 999.978000. CPU Frequency is not max.
 65536      88278            0.00               2206.92            0.035311
Conflicting CPU frequency values detected: 3630.642000 != 1000.251000. CPU Frequency is not max.
 65536      88279            0.00               2206.95            0.035311
Conflicting CPU frequency values detected: 3630.555000 != 1000.389000. CPU Frequency is not max.
 65536      88279            0.00               2206.94            0.035311
Conflicting CPU frequency values detected: 3630.203000 != 1000.201000. CPU Frequency is not max.
 65536      88278            0.00               2206.94            0.035311
Conflicting CPU frequency values detected: 3630.333000 != 1000.147000. CPU Frequency is not max.
 65536      88279            0.00               2206.94            0.035311
Conflicting CPU frequency values detected: 3630.638000 != 999.735000. CPU Frequency is not max.
 65536      88279            0.00               2206.96            0.035311
Conflicting CPU frequency values detected: 3630.562000 != 1001.072000. CPU Frequency is not max.
 65536      88279            0.00               2206.94            0.035311
Conflicting CPU frequency values detected: 3631.155000 != 1000.742000. CPU Frequency is not max.
 65536      88279            0.00               2206.94            0.035311
Conflicting CPU frequency values detected: 3630.251000 != 1000.163000. CPU Frequency is not max.
 65536      88279            0.00               2206.94            0.035311
Conflicting CPU frequency values detected: 3615.712000 != 1000.438000. CPU Frequency is not max.
 65536      88279            0.00               2206.93            0.035311
Conflicting CPU frequency values detected: 3629.303000 != 1000.518000. CPU Frequency is not max.
 65536      88279            0.00               2206.95            0.035311
Conflicting CPU frequency values detected: 3628.639000 != 1000.116000. CPU Frequency is not max.
 65536      88278            0.00               2206.92            0.035311
Conflicting CPU frequency values detected: 3614.048000 != 999.612000. CPU Frequency is not max.
 65536      88264            0.00               2206.56            0.035305
Conflicting CPU frequency values detected: 3630.789000 != 1000.194000. CPU Frequency is not max.
 65536      88279            0.00               2206.97            0.035311
Conflicting CPU frequency values detected: 3630.870000 != 1000.566000. CPU Frequency is not max.
 65536      88278            0.00               2206.93            0.035311
Conflicting CPU frequency values detected: 3630.555000 != 1001.133000. CPU Frequency is not max.
 65536      88279            0.00               2206.95            0.035311
Conflicting CPU frequency values detected: 3630.049000 != 1000.548000. CPU Frequency is not max.
 65536      88279            0.00               2206.92            0.035311
Conflicting CPU frequency values detected: 3629.859000 != 1000.204000. CPU Frequency is not max.
 65536      35058            0.00               876.44             0.014023

 sudo ib_write_bw --rdma_cm --bidirectional --all --ib-dev=mlx5_1 --run_infinitely 192.168.35.11
---------------------------------------------------------------------------------------
                    RDMA_Write Bidirectional BW Test
 Dual-port       : OFF          Device         : mlx5_1
 Number of qps   : 1            Transport type : IB
 Connection type : RC           Using SRQ      : OFF
 PCIe relax order: ON
 TX depth        : 128
 CQ Moderation   : 1
 Mtu             : 1024[B]
 Link type       : Ethernet
 GID index       : 2
 Max inline data : 0[B]
 rdma_cm QPs     : ON
 Data ex. method : rdma_cm
---------------------------------------------------------------------------------------
 local address: LID 0000 QPN 0x1306 PSN 0x960375
 GID: 00:00:00:00:00:00:00:00:00:00:255:255:192:168:35:155
 remote address: LID 0000 QPN 0x57a8 PSN 0x9e92eb
 GID: 00:00:00:00:00:00:00:00:00:00:255:255:192:168:35:11
---------------------------------------------------------------------------------------
 #bytes     #iterations    BW peak[MB/sec]    BW average[MB/sec]   MsgRate[Mpps]
Conflicting CPU frequency values detected: 1197.596000 != 2224.868000. CPU Frequency is not max.
 65536      26               0.00               0.65               0.000010
Conflicting CPU frequency values detected: 1197.585000 != 2237.927000. CPU Frequency is not max.
 65536      13               0.00               0.32               0.000005
Conflicting CPU frequency values detected: 1197.457000 != 3211.183000. CPU Frequency is not max.
 65536      13               0.00               0.32               0.000005
Conflicting CPU frequency values detected: 1197.635000 != 2566.818000. CPU Frequency is not max.
 65536      26               0.00               0.65               0.000010
Conflicting CPU frequency values detected: 1197.671000 != 2037.599000. CPU Frequency is not max.
 65536      13               0.00               0.33               0.000005
Conflicting CPU frequency values detected: 1197.683000 != 1379.349000. CPU Frequency is not max.
 65536      12               0.00               0.30               0.000005
Conflicting CPU frequency values detected: 1197.641000 != 1374.020000. CPU Frequency is not max.
 65536      13               0.00               0.32               0.000005
Conflicting CPU frequency values detected: 1197.640000 != 1297.973000. CPU Frequency is not max.
 65536      26               0.00               0.65               0.000010
Conflicting CPU frequency values detected: 1197.625000 != 1296.392000. CPU Frequency is not max.
 65536      13               0.00               0.33               0.000005
Conflicting CPU frequency values detected: 1197.623000 != 1400.383000. CPU Frequency is not max.
 65536      13               0.00               0.32               0.000005
Conflicting CPU frequency values detected: 1197.714000 != 1456.830000. CPU Frequency is not max.
 65536      13               0.00               0.32               0.000005
Conflicting CPU frequency values detected: 1197.668000 != 2889.561000. CPU Frequency is not max.
 65536      13               0.00               0.33               0.000005
Conflicting CPU frequency values detected: 1197.672000 != 2872.974000. CPU Frequency is not max.
 65536      26               0.00               0.65               0.000010
Conflicting CPU frequency values detected: 1197.588000 != 2867.373000. CPU Frequency is not max.
 65536      13               0.00               0.32               0.000005
Conflicting CPU frequency values detected: 1197.685000 != 2902.293000. CPU Frequency is not max.
 65536      13               0.00               0.32               0.000005
Conflicting CPU frequency values detected: 1197.688000 != 2868.021000. CPU Frequency is not max.
 65536      13               0.00               0.32               0.000005
Conflicting CPU frequency values detected: 1197.686000 != 2868.766000. CPU Frequency is not max.
 65536      26               0.00               0.65               0.000010
Conflicting CPU frequency values detected: 1197.608000 != 2868.979000. CPU Frequency is not max.
 65536      13               0.00               0.32               0.000005
Conflicting CPU frequency values detected: 1197.666000 != 2917.078000. CPU Frequency is not max.
 65536      13               0.00               0.32               0.000005
Conflicting CPU frequency values detected: 1197.621000 != 2901.552000. CPU Frequency is not max.
 65536      13               0.00               0.32               0.000005
Conflicting CPU frequency values detected: 1197.680000 != 1258.628000. CPU Frequency is not max.
 65536      13               0.00               0.32               0.000005
^C


```

**Comment by AlekseyMarchuk** [comment]
Hi @yidong0635 sorry for long delay. Could you try to reproduce the issue using the latest SPDK master branch? I believe that fix for #1292 could fix this issue as well.

**Comment by yidong0635** [comment]
@AlekseyMarchuk , Sorry, since our environments have been damaged, I'm fixing these , It will be fixed later today. But it may not good, I had verified these several days ago. And I will verified these again once all the environments are ok.
Note: This issue hits errors in rdma.c not in nvme_rdma.c. and we use Malloc bdevs.

**Comment by AlekseyMarchuk** [comment]
Yes, I understand that the issue is in rdma target, but most probably it is caused by some problem in initiator

**Comment by yidong0635** [comment]
Let's set this issue as a low level. What you said is reasonable.

**Comment by yidong0635** [comment]
@AlekseyMarchuk  this may not due to the problems at initiator.
from the log, we can see that these error happens after connection/discover finished, and only change 
bs to a 1M other  than 4096, which means that a big size of file for writing could cause error. 
+#define MAX_COMPLETIONS_PER_POLL 128
+
 static int
 spdk_nvmf_rdma_poller_poll(struct spdk_nvmf_rdma_transport *rtransport,
                           struct spdk_nvmf_rdma_poller *rpoller)
 {
-       struct ibv_wc wc[32];
+       struct ibv_wc wc[MAX_COMPLETIONS_PER_POLL];
        struct spdk_nvmf_rdma_wr        *rdma_wr;
        struct spdk_nvmf_rdma_request   *rdma_req;
        struct spdk_nvmf_rdma_recv      *rdma_recv;
@@ -3766,7 +3768,7 @@ spdk_nvmf_rdma_poller_poll(struct spdk_nvmf_rdma_transport *rtransport,
        uint64_t poll_tsc = spdk_get_ticks();

        /* Poll for completing operations. */
-       reaped = ibv_poll_cq(rpoller->cq, 32, wc);
+       reaped = ibv_poll_cq(rpoller->cq, MAX_COMPLETIONS_PER_POLL, wc); // do we need to change this similar as nvme_rdma.c?
        if (reaped < 0) {
                SPDK_ERRLOG("Error polling CQ! (%d): %s\n",
                            errno, spdk_strerror(errno));
some solutions in nvme_rdma.c should add to rdma.c too? after some changes , I confirm that is the code's defect.

**Comment by AlekseyMarchuk** [comment]
Hi @yidong0635, a lot of time passed since last update. Will you be able to reproduce this issue with this patch https://review.spdk.io/gerrit/c/spdk/spdk/+/3736? I believe that old fixes should eliminate core dumps reported below. This patch removes redundant error logs in rdma module.

**Comment by yidong0635** [comment]
hi @AlekseyMarchuk ,
    I'm still on setting up the test environment(It's easily broken), I will verify your patch once ready.

**Comment by yidong0635** [comment]
@AlekseyMarchuk  test finished， and there's no error ， your patch looks good。

**Comment by AlekseyMarchuk** [comment]
Thanks @yidong0635. I'll close this issue when the patch is merged

**Comment by AlekseyMarchuk** [comment]
Patch https://review.spdk.io/gerrit/c/spdk/spdk/+/3736 merged, closing this issue
